crates/reth/statediff/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+
+impl From<BundleState> for BlockStateDiff {
+    fn from(value: BundleState) -> Self {
+        Self {
+            state: value.state,
+            contracts: value.contracts,
+        }
+    }
+}
+
+/// A representation of several [`BlockStateDiff`] ready to be serialized and posted on DA.
+///
+/// N.B. "original" counterparts are needed to correctly construct the [`BatchStateDiff`]
+/// for the range of blocks and not include changes for the keys whose values
+/// were changed inside batch, but end up having the same value as before the batch.
+#[derive(Clone, Default, Debug, Deserialize, Serialize)]
+pub struct BatchStateDiff {
+    /// An account representation in the diff.
+    /// [`Option::None`] indicates the account was destructed, but existed before the batch.
+    pub accounts: HashMap<Address, Option<AccountInfo>>,
+
+    /// The *FIRST* original [`AccountInfo`] for the account in the batch.
+    /// [`Option::None`] if the account was absent before the batch.
+    #[serde(skip)]
+    pub original_accounts: HashMap<Address, Option<AccountInfo>>,
+
+    /// A collection of deployed smart contracts within the batch.
+    pub contracts: HashMap<B256, Bytecode>,
+
+    /// Storage slots.
+    pub storage_slots: HashMap<Address, HashMap<U256, U256>>,
+    /// The *FIRST* original value for the slot in the batch.
+    #[serde(skip)]
+    pub original_storage_slots: HashMap<Address, HashMap<U256, U256>>,
+}
+
+impl BatchStateDiff {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn apply(&mut self, diff: BlockStateDiff) {
+        // Accumulate deployed contracts.
+        for (addr, bytecode) in diff.contracts.into_iter() {
+            self.contracts.insert(addr, bytecode);
+        }
+
+        for (addr, acc) in &diff.state {
+            // Update original account if seen for the first time in the batch.
+            if !self.original_accounts.contains_key(addr) {
+                self.original_accounts
+                    .insert(*addr, acc.original_info.clone());
+            }
+
+            // Now, modify the actual account entry.
+            let cur_account = acc.account_info();
+            if &cur_account == self.original_accounts.get(addr).unwrap() {
+                // Remove if the actual account equals to the original.
+                self.accounts.remove(addr);
+            } else {
+                // The current account is different, update it.
+                self.accounts.insert(*addr, cur_account);
+            }
+
+            let original_acc_storage = self.original_storage_slots.entry(*addr).or_default();
+            let cur_acc_storage = self.storage_slots.entry(*addr).or_default();
+
+            for (key, value) in &acc.storage {
+                // Update original account storage if seen for the first time in the batch.
+                if !original_acc_storage.contains_key(key) {
+                    original_acc_storage.insert(*key, value.original_value());
+                }
```

Comment:
There's this weirdness here when we merge these diffs because we have to carry along the extra originals information inside the diff.  What happens if the two diffs have different originals?  That would probably be a bug, but this feels weird to me.

Reply:
I mean, this weirdness, IIUC, is imposed by the exex interface. If that happens - it pretty much means there's a bug in there.